plugin/src/main/java/org/gradle/testretry/internal/config/TestTaskConfigurer.java

```diff
@@ -46,9 +48,53 @@ public static void configureTestTask(Test test, ObjectFactory objectFactory, Pro
 
         test.getInputs().property("retry.failOnPassedAfterRetry", adapter.getFailOnPassedAfterRetryInput());
 
+        Provider<Boolean> isDeactivatedByTestDistributionPlugin =
+            shouldTestRetryPluginBeDeactivated(test, objectFactory, providerFactory, gradleVersion);
+        test.getInputs().property("isDeactivatedByTestDistributionPlugin", isDeactivatedByTestDistributionPlugin);
+
         test.getExtensions().add(TestRetryTaskExtension.class, TestRetryTaskExtension.NAME, extension);
-        test.doFirst(new ConditionalTaskAction(new InitTaskAction(adapter, objectFactory)));
-        test.doLast(new ConditionalTaskAction(new FinalizeTaskAction()));
+
+        test.doFirst(new ConditionalTaskAction(isDeactivatedByTestDistributionPlugin, new InitTaskAction(adapter, objectFactory)));
+        test.doLast(new ConditionalTaskAction(isDeactivatedByTestDistributionPlugin, new FinalizeTaskAction()));
+    }
+
+    private static Provider<Boolean> shouldTestRetryPluginBeDeactivated(
+        Test test,
+        ObjectFactory objectFactory,
+        ProviderFactory providerFactory,
+        VersionNumber gradleVersion
+    ) {
+        Provider<Boolean> result = providerFactory.provider(() -> callShouldTestRetryPluginBeDeactivated(test));
+        if (supportsPropertyConventions(gradleVersion)) {
+            Property<Boolean> property = objectFactory.property(Boolean.class).convention(result);
+            if (supportsFinalizeValueOnRead(gradleVersion)) {
+                property.finalizeValueOnRead();
```

Comment:
memoize

Reply:
❓ Isn't there way to have a `PropertyWithFinalizedValueOnRead` from the very beginning? I can see how flexible `finalizeValueOnRead()` can be. But at the same time this also looks like you can get it wrong...